### DISCUSSION TITLE
Rewriting assertions in closure functions (i.e. function nested in another function)

### DISCUSSION THREAD

**Discussion by mboutet** [pytest-dev/pytest#8679] [General] [answered]
I have a test helper which I implemented using a fixture so that this helper can also use other fixtures (`make_foo` and `make_bar`in my example below):

```python
# in conftest.py
@pytest.fixture(scope="function")
def helper(make_foo, make_bar) -> callable:
    def _helper(param1: str, param2: str):
        assert False

    return _helper


# in test_foo.py
@pytest.mark.parametrize("param1", ["a", "b"])
@pytest.mark.parametrize("param2", ["c", "d"])
def test_foo(helper, param1, param2):
    helper(param1, param2)
```

This is of course a dummy example as my helper uses more than 10 fixtures.

The problem is that even though I register `pytest.register_assert_rewrite("tests.conftest")` in my root `__init__.py`, pytest does not rewrite the assertions inside `def _helper`.

Is rewriting assertions in closure functions supported by pytest?

**Comment by mboutet** [comment, marked as answer]
Ok, nevermind.

> Rewritten assert statements put introspection information into the assertion failure message. pytest only rewrites test modules directly discovered by its test collection process, so asserts in supporting modules which are not themselves test modules will not be rewritten.

**Reply by RonnyPfannschmidt** [reply]
as far as i recall all pytest plugins and conftest files should get rewritten as well
